src/super_gradients/training/datasets/collate_fn.py

```diff
@@ -0,0 +1,421 @@
+from torch.utils.data.dataloader import default_collate
+import random
+from typing import List, Union, Tuple, Optional, Dict, Type, Callable, Iterable
+
+from abc import ABC
+
+import cv2
+import torch
+import numpy as np
+
+
+from data_gradients.dataset_adapters.base_adapter import BaseDatasetAdapter
+from data_gradients.dataset_adapters.detection_adapter import DetectionDatasetAdapter
+from data_gradients.dataset_adapters.segmentation_adapter import SegmentationDatasetAdapter
+from data_gradients.dataset_adapters.classification_adapter import ClassificationDatasetAdapter
+
+from super_gradients.common.registry.registry import register_collate_function
+from super_gradients.common.decorators.factory_decorator import resolve_param
+from super_gradients.common.factories.collate_functions_factory import CollateFunctionsFactory
+from super_gradients.training.utils.detection_utils import xyxy2cxcywh
+
+from super_gradients.common.environment.device_utils import device_config
+from super_gradients.common.environment.ddp_utils import get_local_rank
+from super_gradients.training.utils.distributed_training_utils import wait_for_the_master
+from super_gradients.training.datasets.segmentation_datasets.segmentation_dataset import SegmentationDataSet
+
+
+class DatasetItemsException(Exception):
+    def __init__(self, data_sample: Tuple, collate_type: Type, expected_item_names: Tuple):
+        """
+        :param data_sample: item(s) returned by a dataset
+        :param collate_type: type of the collate that caused the exception
+        :param expected_item_names: tuple of names of items that are expected by the collate to be returned from the dataset
+        """
+        collate_type_name = collate_type.__name__
+        num_sample_items = len(data_sample) if isinstance(data_sample, tuple) else 1
+        error_msg = f"`{collate_type_name}` only supports Datasets that return a tuple {expected_item_names}, but got a tuple of len={num_sample_items}"
+        super().__init__(error_msg)
+
+
+class BaseDatasetAdapterCollateFN(ABC):
```

Comment:
Let's split this file and have one file per class.
So instead of `src/super_gradients/training/datasets/collate_fn.py` there would be
```
src/super_gradients/training/datasets/collate_fn/__init__.py
src/super_gradients/training/datasets/collate_fn/ppyolo_collate_fn.py
src/super_gradients/training/datasets/collate_fn/ppyolo_crowd_collate_fn.py
...
```

Reason: 
One file easier to digest
Easier to do diff between two files and see their difference.

Reply:
Fair enough, I'll do that

Reply:
I did it here https://github.com/Deci-AI/super-gradients/pull/1470
Eventually, I gave up on moving it in the `dataset` folder, I just lost more than 1h of trying to fix circular import loops and I think it is just too deep...